SP23.P03.Web/Controllers/TrainsController.cs

```diff
@@ -0,0 +1,153 @@
+            }
+
+            var trainDto = new TrainDto
+            {
+                Id = train.Id,
+                Name = train.Name,
+                Status = train.Status,
+                Capacity = train.Capacity
+            };
+
+            return Ok(trainDto);
+        }//end GetTrainById
+
+        [HttpPost]
+        [Authorize(Roles = RoleNames.Admin)]
+        public ActionResult<CreateTrainDto> CreateTrain(CreateTrainDto createTrainDto, int id)
+        {
+            var train = trains.FirstOrDefault(x => x.Id == id);
+
+            if (InvalidCreateTrainDto(createTrainDto))
+            {
+                return BadRequest();
+            }
+
+            var createdTrain = new Train
+            {
+                Name = createTrainDto.Name,
+                Status = createTrainDto.Status,
+                Capacity = createTrainDto.Capacity
+            };
+
+            trains.Add(createdTrain);
+            dataContext.SaveChanges();
+
+            var trainDto = new TrainDto {
+                Id = createdTrain.Id,
+                Name = createdTrain.Name,
+                Status = createdTrain.Status,
+                Capacity = createdTrain.Capacity
+            };
+
+            return Ok(trainDto);
+        }//end CreateStation
+
+        [HttpPut("{id}")]
+        [Authorize(Roles = RoleNames.Admin)]
+        public ActionResult<TrainDto> UpdateTrain([FromBody]CreateTrainDto createTrainDto,
+                                                  [FromRoute] int id)
+        {
+            var train = trains.FirstOrDefault(x => x.Id == id);
+
+            if (train == null)
+            {
+                return NotFound();
+            }
+
+            if (InvalidCreateTrainDto(createTrainDto))
+            {
+                return BadRequest();
+            }
+
+            train.Name = createTrainDto.Name;
+            train.Status = createTrainDto.Status;
+            train.Capacity = createTrainDto.Capacity;
+
+            dataContext.SaveChanges();
+
+            var trainDto = new TrainDto
+            {
+                Id = train.Id,
+                Name = train.Name,
+                Status = train.Status,
+                Capacity = train.Capacity
+            };
+
+            return Ok(trainDto);
+        }//end UpdateTrain
+
+        [HttpDelete("{id}")]
+        [Authorize(Roles = RoleNames.Admin)]
+        public ActionResult DeleteTrain(int id, TrainDto dto)
+        {
+            var train = trains.FirstOrDefault(x => x.Id == id);
+
+            if (train == null)
+            {
+                return NotFound();
+            }
+
+            trains.Remove(train);
+
+            dataContext.SaveChanges();
+
+            return Ok();
+        }
+
+        private bool InvalidCreateTrainDto(CreateTrainDto createTrainDto) =>
+           createTrainDto == null
+           || String.IsNullOrEmpty(createTrainDto.Name)
+           || String.IsNullOrEmpty(createTrainDto.Status);
```

Comment:
```suggestion
        private bool InvalidCreateTrainDto(CreateTrainDto createTrainDto) =>
           createTrainDto == null
           || String.IsNullOrWhiteSpace(createTrainDto.Name)
           || String.IsNullOrWhiteSpace(createTrainDto.Status)
           || createTrainDto.Capacity > 0;
```
1. I prefer `String.IsNullOrWhiteSpace` because it also prevents names that are just whitespace like " "
2. Capacity check